### ISSUE TITLE
MGLShapeSource performance issues with high quantity of simple polygons

### ISSUE THREAD

**Issue by MattCarr147** [mapbox/mapbox-gl-native#14458]
Firstly apologies if this is more of a query than a bug, however I've tried looking through the documentation and couldn't find any clarification for upper limits on performance etc. 

I'm attempting to display roughly 3,500 polygons (~610 metre squares) using an MGLShapeSource. I retrieve the data from a database and construct a number of `MGLPolygonFeature` instances. The source is updated as follows:

```
self.shapeSource.shape = MGLShapeCollectionFeature(shapes: clusters)
```

I've disabled geometry simplification of the source as I found that the grid squares were being removed entirely at low zoom levels which was undesirable. 

The map takes a few moments to render the data but the significant performance issues begin when zooming and panning the map, as the map frequently stutters and freezes. Is this to be expected? I noticed that other issues have been raised regarding performance when updating styles in sources containing large numbers of features but in my case once the initial load is complete its data & styling doesn't change.

I'm happy to accept it might simply be a case of trying to render excessive amounts of data, but as mentioned there's no indication of what the upper limit would be in the docs. So if this could be clarified and the reasons why that may be helpful going forwards?

Also note that I'm using data driven styling for the layer as follows:

- Squares are coloured green, orange or blue depending on a string status
- Opacity changes depending on zoom level (the closer you zoom in the less opaque)

I've just tried disabling all these styles and painting each square a solid blue colour and the performance issues remain.

### Steps to reproduce

 1. Initialise an MGLShapeSource containing a large number of simple polygons (squares)
 2. Pan and zoom the map

### Expected behavior

The map can be panned and zoomed without any noticeable performance issues.

### Actual behavior

The map 'stutters' and freezes occasionally when navigating the map, particularly when zooming. 

### Configuration

**Mapbox SDK versions: 4.9.0**
**iOS/macOS versions: 12.2**
**Device/simulator models: iPad Mini 4th Generation**
**Xcode version: 10.2 (beta 3)**

**Comment by jmkiley** [comment]
Hello @MattCarr147 -

Thank you for your report. Could you share how you are adding the `MGLPolygonFeature` objects to your map? Are you adding them to the map directly, or using the `MGLShapeSource` to create a [`MGLFillStyleLayer`](https://docs.mapbox.com/ios/api/maps/4.10.0/Classes/MGLFillStyleLayer.html)?

**Comment by MattCarr147** [comment]
Hi, sorry for the delay in getting back to you. 

Just to confirm that I'm using an `MGLFillStyleLayer` and adding/setting the features as described in the original report.

**Comment by MattCarr147** [comment]
Apologies for the bump on this issue. I noticed it's flagged with 'needs information' and I'm not too sure whether I'm expected to be providing it? Though as it happens, I do have a little bit of information to add which may help. 

As a workaround for this issue, I now display circles instead of polygons at low zoom levels. To achieve this I add an MGLPointFeature per MGLPolygonFeature to my ShapeSource (so it now contains ~7,000 features) and use runtime styling to only render circle features at low zoom levels and then polygons after a certain point (e.g. zoom level 11+).

I found this approach improved the user experience when panning/zooming, maybe suggesting that the issue is caused by trying to render too many vertices at once?
